auto_detect.py

```diff
@@ -0,0 +1,162 @@
+            constraint_name = file.read().strip()
+            if constraint_name != 'long_term':
+                raise RuntimeError(f"Constraint /sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_name was not long_term, but {constraint_name}")
+
+        file_path = '/sys/class/powercap/intel-rapl/intel-rapl:0/constraint_0_max_power_uw'
+        with open(file_path, 'r', encoding='UTF-8') as file:
+            tdp = file.read()
+            data['tdp'] = int(tdp) / 1_000_000
+
+        logger.info('Found TDP: %d W', data['tdp'])
+    #pylint: disable=broad-except
+    except Exception as err:
+        logger.info('Exception: %s', err)
+        logger.info('Could not read RAPL powercapping info from /sys/class/powercap/intel-rapl')
+
+    try:
+        file_paths = {
+            1: '/sys/class/powercap/intel-rapl/intel-rapl:0/name',
+            2: '/sys/class/powercap/intel-rapl/intel-rapl:1/name',
+            3: '/sys/class/powercap/intel-rapl/intel-rapl:2/name',
+            4: '/sys/class/powercap/intel-rapl/intel-rapl:3/name',
+            5: '/sys/class/powercap/intel-rapl/intel-rapl:4/name',
+            6: '/sys/class/powercap/intel-rapl/intel-rapl:5/name',
+        }
+        for chips, file_path in file_paths.items():
+            with open(file_path, 'r', encoding='UTF-8') as file:
+                domain_name = file.read().strip()
+                if domain_name != f"package-{chips-1}":
+                    raise RuntimeError(f"Domain {file_path} was not package-{chips-1}, but {domain_name}")
+                logger.info('Found Sockets: %d', chips)
+                data['chips'] = chips
+    #pylint: disable=broad-except
+    except Exception as err:
+        logger.info('Exception: %s', err)
+        logger.info('Could not find (additional) chips info under file path. Most likely reached final chip. continuing ...')
+
+
+    try:
+        cpuinfo = subprocess.check_output('lscpu', encoding='UTF-8')
+        match = re.search(r'On-line CPU\(s\) list:\s*(0-)?(\d+)', cpuinfo)
+        if match:
+            data['threads'] = int(match.group(2))+1 # +1 because 0 indexed
+            logger.info('Found Threads: %d', data['threads'])
+        else:
+            logger.info('Could not find Threads. Setting to None')
+
+        # this will overwrite info we have from RAPL socket discovery, as we
+        # deem lscpu more relieable
+        match = re.search(r'Socket\(s\):\s*(\d+)', cpuinfo)
+        if match:
+            data['chips'] = int(match.group(1))
+            logger.info('Found Sockets: %d (will take precedence if not 0)', data['chips'])
+        else:
+            logger.info('Could not find Chips/Sockets via lscpu')
+
+        if data['threads'] and data['chips']:
+            match = re.search(r'Thread\(s\) per core:\s*(\d+)', cpuinfo)
+            if match:
+                threads_per_core = int(match.group(1))
+                data['cores'] = (data['threads'] // threads_per_core) // data['chips']
```

Comment:
I think the problem is here that not every core has 2 threads on my computer. You could get the cores through getting 
```
    cores_per_socket_match = re.search(r'Core\(s\) per socket:\s*(\d+)', input_str)
    sockets_match = re.search(r'Socket\(s\):\s*(\d+)', input_str)


    if cores_per_socket_match and sockets_match:
        cores_per_socket = int(cores_per_socket_match.group(1))
        sockets = int(sockets_match.group(1))
        data['cores'] cores_per_socket * sockets
```

Reply:
Can you please post the full dumps of `lscpu`, `hwinfo` and `/proc/cpuinfo`.

I cannot fully implement the regex without it.

Reply:
[commands.txt](https://github.com/green-coding-berlin/spec-power-model/files/13808787/commands.txt)

Reply:
@ArneTR ping